### ISSUE TITLE
Middle-click clipboard in RStudio

### ISSUE THREAD

**Issue by adatum** [rstudio/rstudio#1667]
As a linux user, I like highlighting text to copy, and middle-clicking to paste, as it is fast. I often wish to do this between the editor and console in RStudio, but it does not work. Is there any possibility of fixing this in the future?

**Comment by kevinushey** [comment]
This seems like an unfortunate omission on our part -- we did some work to enable middle-click pasting in v1.1, and this does work (RStudio -> other application), but unfortunately middle-click pasting doesn't seem to work in RStudio editors.

**Comment by adatum** [comment]
You're right, middle-click pasting works from RStudio to other applications, but not within RStudio, or from other applications to RStudio.

If you can get this working it will be hugely appreciated and streamline the workflow in RStudio!

**Comment by kevinushey** [comment]
Playing around a little bit more, it seems like there is something a bit trickier going on -- it seems like it works half the time, and doesn't work the other half of the time. The following consistently works for me:

1. Enter some text into an editor in the Source pane,
2. Press <kbd>Ctrl + A</kbd> to highlight the text,
3. Middle-click somewhere on the console.

Those middle-clicks seem to succeed consistently. (I did learn that one does need to middle-click quickly -- if you press the middle button down, hold it for a second, and then release, the middle-click paste does not occur. This appears to be a universal behavior; not specific to RStudio.)

However, the same action from Console -> Source editor does not work until the second click. The first click focuses the editor but loses the paste; the second click successfully pastes.

**Comment by adatum** [comment]
Not sure, but maybe it's a focus issue. 

Sometimes highlight/middle-click works perfectly between external applications/Console/Editor, other times it doesn't work at all or just in certain directions. 

Sometimes it pastes directly where the mouse pointer was when it was middle-clicked (expected behavior), other times it pastes where there is an existing cursor blinking (independent of the mouse pointer) in the editor.  

I haven't found a pattern yet.

System info: 
RStudio 1.1.383
R version 3.4.2 (2017-09-28)
Platform: i686-pc-linux-gnu (32-bit)
Running under: Ubuntu 16.04.3 LTS

**Comment by kevinushey** [comment]
For what it's worth, the behavior seems a lot more consistent (at least for me) running with Ubuntu 17.10 (as that now uses Gnome 3); it's possible that we're running afoul of some Unity-specific misbehaviors here.

**Comment by adatum** [comment]
I don't use Unity. Gnome2/MATE.

**Comment by kevinushey** [comment]
Thanks for clarifying. We'll see if we can get to the bottom of this.

**Comment by lwhansen** [comment]
Just want to reiterate, that this is still not working with latest version (1.1.383). I consistently get the old selection on first middle click and the new on second click. This is within a notebook window, running latest Kubuntu. Quite frustrating.

**Comment by ClaytonJY** [comment]
As an extra data point, I'm also using RStudio v1.1.383, but on Linux Mint v18.3 (based on Ubuntu 16.04) w/ KDE 5, and middle-clicking works both to and from RStudio.

I don't know if this could be it, but @lwhansen are you on an older KDE version? `kf5-config --version` says I'm on KDE 5.36.

**Comment by lwhansen** [comment]
I am actually quite bleeding-edge. Maybe that's the problem? On Ubuntu 17.10 running KDE 5.38.0. Downloading RStudio 1.1.442 now. 

I do run all kids of other programs. Typically several different browsers, an IDE, emacs, slack and much more. Maybe I will do a reboot and see when the problems starts happening. Thanks

**Comment by KBGrammer** [comment]
I'm on Ubuntu 16.04 running Unity 7.4.5. Middle click paste is hit or miss (mostly miss) for me.

I tried highlighting in one R script and then middle click pasting into another, and it pastes nothing.

I tried highlighting a comment from this thread and then middle click pasting into an R script, and it does nothing.

I tried to a comment from this thread, pressing ctrl+c, and then middle click pasting it into an R script, and it pastes what I copied. That's still not the "highlight, middle click to paste" functionality that I'm expecting, though.

I tried highlighting something new, pressing ctrl+c, and then middle click pasting into an R script, and it pastes the same thing that it pasted from before.

**Comment by adrfantini** [comment]
I am on Linux Mint 18 KDE + Rstudio version 1.0.143 and select-to-copy continues to behave very strangely, as if it is reading and writing to a weird clipboard (not KDE's clipboard).

Selecting text in the Rstudio console or from Rstudio menus does not save it to the clipboard, and center-click does not paste it on the Rstudio console. Weirdly enough, it DOES paste the selected text on a terminal. My guess is Rstudio is talking to X's clipboard, but not to KDE's?

**Comment by lwhansen** [comment]
I have update everything and rebooted. That doesn't fix the problem for me. They only reliable way that copy-n-paste works for me is Ctrl+C and Ctrl+V. Definitely not ideal under Linux, where most people just use middle click. I think this is a GWT limitation.

**Comment by wc18** [comment]
Exactly same issue as lwhansen running Ubuntu 16.04 with Gnome. It's a real pain to use RStudio like this.

**Comment by KBGrammer** [comment]
It inconsistently middle-click pastes things that I highlight in RStudio and paste in RStudio, from outside RStudio into RStudio, and from RStudio to outside RStudio. It is also somehow extremely sensitive in how I highlight code in RStudio, and if I try to highlight a line or two to delete them but move my mouse a little too far it behaves as if I intended to highlight the entire script and I end up deleting the whole script and having to ctrl+Z to undo it because I didn't realize that moving the mouse 10 pixels too far highlighted the whole document.

It also doesn't make sense that if I want to replace instances of some text in a selected area, there's a intricate song and dance routine that determines the order in which I have to open the find&replace menu, select the area, and press the "replace in selection" check box, and it fails if I don't do it in the right sequence. For instance, if I highlight text, click "edit", and click "find and replace", nothing happens. If I click off of text and then click edit and find and replace, the options show up. If I start typing what I want to find, it searches and then I have to navigate back to where I started (annoying). If I highlight and then start typing in the find box, it searches the whole document and I have to navigate back to where I started (annoying). If I click "in selection" and then go highlight something, it unchecks "in selection" (annoying). Only if I start unhighlighted, open the find and replace menu, highlight, and click "in selection" does it behave as I expect it to and it only works if I do the song and dance in that order. It's infuriating.

It doesn't make any sense that I can highlight 2 lines of a 100 line script and it highlights all of them, and the inconsistent behavior of middle-click paste is terrible.

It makes RStudio borderline unusable, quite frankly.

**Comment by kevinushey** [comment]
For everyone encountering issues with middle-pasting: can you let us know if things are any better in the RStudio [preview release](https://www.rstudio.com/products/rstudio/download/preview/)?

@KBGrammer: thanks for the bug report re: Find and Replace; I'll file that as a separate issue but let's keep this issue thread scoped to the middle paste issues.

**Comment by lwhansen** [comment]
Looks better, but I need centos tarball to really test.

**Comment by KBGrammer** [comment]
Sorry for that, I kind of tied them together since they are all text highlighting problems.

I just updated to the current release version (1.1.453), I'm not sure what exact version I had previously but it was recent (1.1.4XX). If I highlight text in RStudio and then middle click in RStudio, nothing happens. If I middle click in emacs, it pastes what I highlighted. If I highlight what I pasted into emacs and try to paste it into RStudio, nothing happens.

I then tried 1.2.747 for Ubuntu. Middle click paste for RStudio->Rstudio, RStudio -> emacs, and emacs -> RStudio all works.

**Comment by adatum** [comment]
I switched to Fedora 28 MATE now, and with RStudio 1.1.453, highlight + middle-click paste is working. :)

I tested all permutations between the RStudio editor, RStudio console, and an editor/terminal outside of RStudio.

Also, the pasting is exactly at the location of the mouse cursor when middle-click pasting, which is the expected and desired behavior. (as opposed to pasting to the location of the blinking cursor in the RStudio editor)

**Comment by adatum** [comment]
Yes, I thought it was working perfectly but sometimes it still doesn't work properly. 

Sometimes:
- highlighting doesn't copy.
- pasting with middle click doesn't work.
- pasting with middle-click sometimes pastes the highlight-to-copy buffer, and other times the right-click-and-copy (or Ctrl-C) buffer.
- pasting with middle click inserts where the blinking cursor is in the editor instead of where the mouse pointer was middle-clicked.

It's working correctly more often than when the issue was first opened, but it's still inconsistent.

**Comment by jmcphers** [comment]
We don't have a consistent repro for this and haven't received any additional reports in well over a year, so closing. We also have some reports above that RStudio 1.2 (which includes improvements to copy) makes things better, so hopefully that will do the job for most people having trouble! If not, feel free to open a new issue.

**Comment by adrfantini** [comment]
The point of this bug is that the behaviour is completely *inconsistent*, so of course you cannot have a consistent reprex. Ignoring the problem will not fix it, unfortunately.
I am on Ubuntu with Rstudio 1.2.5019 and encounter this more often than not. One thing that happens often, but not always, is:

1. Start Rstudio, do some stuff for a bit
2. HIghlight something
3. Middle click paste. Does nothing. Try again. Nothing. On the third try it usually succeeds.

Right now I can reproduce this reliably, who knows if after an Rstudio restart it will happen again.

This is made even worse by the [known](https://community.rstudio.com/t/double-clicking-to-select-all-in-r-studio/29923) [highlighting](https://support.rstudio.com/hc/en-us/community/posts/213990828-Highlighting-a-word-with-double-click-not-working) issue, where double-clicking a word (or triple-clicking a line) sometimes selects the whole text, like a CTRL+A action. These two issues together make mouse interaction on Rstudio under Linux feel extremely frustrating.

I'm more than happy to try and debug this for you if you tell me what could be useful.

**Comment by assaron** [comment]
I agree completely with @adrfantini : the bug definitely wasn't fixed and is still present, so I don't understand the point of closing the issue. For me in _consistently_ doesn't work on the first try. Only from this thread I discovered, that if you do middle-click multiple times it can paste something on the second or third time.

I would be happy to help if someone can tell what to do.

**Comment by elchudi** [comment]
Also confirming that the bug is still present in Version 1.3.1073 running on Ubuntu 20.04

**Comment by kevinushey** [comment]
Re-opening since the issue doesn't appear to be fixed (at least, not in all configurations).

If you're experiencing issues with middle-click pasting, it would be very helpful to know a bit more about your system environment (what OS; what window manager; anything else that might differ from defaults or make the issue easier or harder to reproduce)

**Comment by elchudi** [comment]
Hey, my configuration is a plain stock Ubuntu 20.04 up to date with what i think is the latest RStudio. I've attached more info. 

wmctrl -m

- Name: GNOME Shell
- Class: N/A
- PID: N/A
- Window manager's "showing the desktop" mode: OFF

lsb_release -a:

- No LSB modules are available.
- Distributor ID:	Ubuntu
- Description:	Ubuntu 20.04.1 LTS
- Release:	20.04
- Codename:	focal

RStudio About:

- Version 1.3.1073
- © 2009-2020 RStudio, PBC
- "Giant Goldenrod" (718e6d75, 2020-07-29) for Ubuntu Bionic
- Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) QtWebEngine/5.12.8 Chrome/69.0.3497.128 Safari/537.36

gnome-shell --version:

- GNOME Shell 3.36.4

**Comment by kevinushey** [comment]
I'm testing in an Ubuntu 20.04 VM (macOS 10.15.7 host; using Parallels).

- In general, middle-click pasting from other applications to RStudio works without issue.

- When I try to middle-click to paste content from RStudio to another application, I often see (for multi-line selections) that only part of the selection is pasted. For example, if I enter <kbd>Ctrl + A</kbd> to select the entire document, middle-clicking seems to only paste the last line.

Can anyone else confirm that the primary issue lies with pasting from RStudio to other applications? Or do you also see issues with pasting from other applications to RStudio?

**Comment by adatum** [comment]
@kevinushey I can confirm what you described: multi-line (eg. 3 line) selections from the RStudio editor to an external application or to RStudio console is middle-click pasting only part of the selection (eg. 2 lines). Ctrl-C Ctrl-V doesn't have that issue. 

So far middle-click pasting  of multi-line selections in other directions seems ok: external applications to RStudio editor or console; RStudio console to RStudio editor or external application. 

I can't guarantee there aren't other issues, but that's what I've observed right now.

RStudio Version 1.3.1093
R version 4.0.2 (2020-06-22)
Platform: x86_64-redhat-linux-gnu (64-bit)
Running under: Fedora 32 (Thirty Two)
MATE desktop environment 
Metacity (Marco) window manager

**Comment by nfultz** [comment]
Drag to copy and middle click to paste works fine for me in the editor and terminal tabs. In the R console copy works but not paste.

###

    RStudio Edition : Server [Open Source]
    RStudio Version : 1.2.1335
    OS Version      : Ubuntu 20.04.1 LTS
    R Version       : R version 4.0.2 (2020-06-22)

GNOME Shell 3.36.4

**Comment by elchudi** [comment]
Some issues still seems to be present. 

Several findings that i can reproduce 100% of the time:
_ I can middle-click paste to the console once. After the first successful paste, the middle-click clipboard is empty and i cannot longer paste, in or outside of rstudio. 
_ I cannot middle-click paste to a script file. 
_ I can middle-click paste to the terminal successfully multiple times. 
_ I can highlight anything in the console, the terminal or a script file and it will be available in the middle-click clipboard to paste


Version: RStudio 2022.12.0+353 "Elsbeth Geranium" Release (7d165dcfc1b6d300eb247738db2c7076234f6ef0, 2022-12-03) for Ubuntu Jammy
Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) rstudio/2022.12.0+353 Chrome/102.0.5005.167 Electron/19.1.3 Safari/537.36

**Comment by marbotte** [comment]
Same problem here in the pre-release build 2023.03.0, middle click does not work in rstudo, from outside to rstudio, from rstudio to outside...
> RStudio 2023.03.0-daily+327 "Cherry Blossom" Daily (547c4aba6f12c2284ead701b505b1ef68413fcd9, 2023-02-18) for Gentoo
> Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) QtWebEngine/5.15.8 Chrome/87.0.4280.144 Safari/537.36
> kernel version: linux-6.1.12-gentoo

Note that I use the vim keybindings...

**Comment by vmkalbskopf** [comment]
6 years on... sigh

**Comment by khal3dj** [comment]
Using 2023.06.0 Build 421 on jammy + Cinnamon 5.2.7. 
Highlighting + middle click will either paste:
1. nothing
2. two first lines
3. last line
4. everything

**Comment by allefeld** [comment]
Still doesn't work with RStudio 2023.12.0 Build 369 on Debian 12 with KDE 5.27.5.

**Comment by jagerman** [comment]
> As a linux user ...

I think this is where they tuned out reading the bug report at any point over the last six and a half years. :disappointed:

**Comment by ronblum** [comment]
I'll mark this for re-review. I can't say when or if we'll fix this, but we'll re-review the bug.

Also, @jagerman I can assure you that Linux wasn't the issue! A bug is still a bug. We do have a number of criteria for deciding on whether or not to address a bug, but with supporting 11 Linux distributions and versions, we can't ignore bugs just because they're found on Linux. I know this well because I'm the one who certifies the builds against the Linux instances. 🙂

**Comment by ronblum** [comment]
Question: Has anyone tried this with the editor in visual mode or only source mode?

**Comment by jagerman** [comment]
Investigating this a little further:
- if I select something in another window and then middle click into the RStudio source editor *without* letting the RStudio window have focus first then it will successfully paste exactly once.
- if I do the same into the Console, then it sometimes works on the first try, sometimes on the second try, and sometimes doesn't work at all.
- if I select, then left-click *on the editor part* of the window to bring RStudio into focus, then middle click it doesn't paste.
- if I select, then left-click *on the title bar* to bring it into focus, then middle click into the editor, it works exactly once.

But I think the real issue is this: *any* operations in the editor and Console -- typing, or even just moving the cursor by left-clicking or using arrow keys -- clears the global selection buffer (i.e. the selected test *in the other application* gets cleared).  Middle-click pasting included -- which is why in even the success cases above, it only works once.  This should not be happening: the global selection buffer should only be updated if I actually select something (or clear a selection), and I suspect if this was fixed, all the other issues would disappear.

**Comment by ronblum** [comment]
@jagerman Thank you for info, including the hint!

**Comment by kevinushey** [comment]
For reference; we did a fair amount of work to get middle-click pasting to work with the older Qt builds of RStudio on Linux; our hope was that middle-click pasting would "just work" out-of-the-box after switching to Electron. We might still need to add some glue code to get middle-click pasting hooked up.

**Comment by petrelharp** [comment]
Some more observations (on current Debian Sid; Rstudio 2024.12.0 Build 467):
- this applies to both console and editor
- the cursor has two states: (1) grey and not flashing, and (2) black and flashing
- middle-click-paste only seems to work when the window (either console or editor) has not just been in focus; when moving between windows within Rstudio, this is indicated by a grey cursor - so at first I thought it only works when the cursor is in state (1) - but as noted by others above, pasting into Rstudio from another application works (but only once) when the cursor is in state (2)
- After middle-click pasting into Rstudio, the middle-click-paste buffer has some weird content in it: middle-click-pasting to the terminal makes the bell sound.
- However, just clicking to focus in some window (which makes middle-click-paste not work in Rstudio) does not clear the buffer: I can select something in Rstudio, left-click to focus a different window, and then middle-click-paste the content elsewhere.
- If I middle-click somewhere in the editor that's not where the cursor currently is that is not the end of a line (and am in a state it would normally paste), then it moves the cursor but does not paste. Doing this does clear the clipboard.

To be clear about this being a serious bug:  suppose I've got text in a terminal that I want to paste into Rstudio's editor. I can't use Ctrl-C because it's the terminal (eg, bash). As noted above, this will only work if the cursor is already where I want the text to go; otherwise it moves the cursor and loses what's in the clipboard. (with some caveats about the end of the line above; sometimes this works and sometimes it doesn't)
